File: frontend/src/api/auditCheck.ts
/**
 * 稽核檢查（Audit Check）— API 封裝
 * 對應後端 /api/v1/audit-check（app/routers/audit_check.py）
 *
 * 規格：docs/SPEC_audit_check.md
 * ⚠️ CLAUDE.md §6：元件內不直接用 axios，一律透過這裡的函式。
 */
import client from './client'

// ── 型別 ──────────────────────────────────────────────────────────────────

/** 判定類型（取代 Excel 的字色語意，由使用者自行維護） */
export interface ResultType {
  id: number
  code: string
  label: string
  color: string
  /** 是否計入「達標項數」 */
  counts_as_pass: boolean
  /** 是否列入「缺失」自動彙整 */
  include_in_summary: boolean
  is_default: boolean
  is_system: boolean
  sort_order: number
  is_active: boolean
}

export interface AuditItem {
  id: number
  parent_id: number | null
  name: string
  description: string | null
  sort_order: number
  is_active: boolean
  /** 已被任一期稽核單引用 → 不可改名／刪除，只能停用 */
  in_use: boolean
  children: AuditItem[]
}

export interface SheetSummary {
  id: number
  company_id: number
  company_name: string
  audited_on: string | null
  status: string
  completion_rate: number | null
  completion_label: string
}

export interface Period {
  id: number
  period: string
  title: string
  goal_major: number
  goal_minor: number
  note: string | null
  /** 覆核區是否計為一個稽核子項（影響各部門分數） */
  review_counts_in_score: boolean
  status: string
  created_at: string
  sheets: SheetSummary[]
}

export interface SheetDepartment {
  id: number
  department_id: number
  name: string
  sort_order: number
  deficiency_override: string | null
  deficiency: string
}

export interface SheetItem {
  id: number
  item_id: number
  parent_sheet_item_id: number | null
  /** 1 = 大項、2 = 子項（只有子項計分） */
  level: number
  display_no: string | null
  name: string
  scope_note: string | null
  target_department_ids: number[]
  sort_order: number
}

export interface Cell {
  sheet_item_id: number
  sheet_department_id: number
  comment: string | null
  result_code: string
  updated_at: string | null
}

export interface DepartmentScore {
  sheet_department_id: number
  department_id: number
  name: string
  sub_count: number
  pass_count: number
  score: number | null
  score_label: string
}

export interface Review {
  sheet_department_id: number
  source_period: string | null
  pending_text: string | null
  pending_result: string
  result_text: string | null
  result_status: string
}

export interface SheetDetail {
  id: number
  period_id: number
  period: string
  title: string
  goal_major: number
  goal_minor: number
  review_counts_in_score: boolean
  company_id: number
  company_name: string
  audited_on: string | null
  audited_session: string | null
  audited_label: string
  executor_label: string | null
  /** 完成率分母的人工調整（Excel 的「3*3-1=8」：本期少做一項填 -1） */
  completion_adjust: number
  remark: string | null
  status: string
  departments: SheetDepartment[]
  items: SheetItem[]
  cells: Cell[]
  scores: DepartmentScore[]
  reviews: Review[]
  completion_rate: number | null
  completion_label: string
  result_types: ResultType[]
}

export interface SheetItemSpec {
  item_id: number
  scope_note?: string | null
  target_department_ids: number[]
}

export interface StatisticsCell {
  period: string
  score: number | null
  label: string
}

export interface StatisticsBlock {
  company_id: number
  company_name: string
  periods: string[]
  audited_labels: string[]
  major_items: string[][]
  rows: { department_id: number; name: string; cells: StatisticsCell[] }[]
  completion: StatisticsCell[]
}

export interface Statistics {
  year: number
  note: string | null
  blocks: StatisticsBlock[]
}

export interface FlaggedRow {
  period: string
  company_name: string
  department_id: number
  department_name: string
  item_name: string
  display_no: string
  result_code: string
  result_label: string
  result_color: string
  comment: string
}

export interface MobileSheetRow {
  id: number
  period: string
  title: string
  company_id: number
  company_name: string
  audited_on: string | null
  status: string
  completion_rate: number | null
  completion_label: string
}

// ── 判定類型 ──────────────────────────────────────────────────────────────
export const resultTypesApi = {
  list: () => client.get<ResultType[]>('/audit-check/result-types'),
  create: (body: Partial<ResultType> & { code: string; label: string }) =>
    client.post<ResultType>('/audit-check/result-types', body),
  update: (id: number, body: Partial<ResultType>) =>
    client.put<ResultType>(`/audit-check/result-types/${id}`, body),
  toggle: (id: number) => client.patch<ResultType>(`/audit-check/result-types/${id}/toggle`),
  remove: (id: number) => client.delete(`/audit-check/result-types/${id}`),
}

// ── 檢查項主檔 ────────────────────────────────────────────────────────────
export const itemsApi = {
  list: (includeInactive = false) =>
    client.get<AuditItem[]>('/audit-check/items', { params: { include_inactive: includeInactive } }),
  create: (body: { name: string; parent_id?: number | null; description?: string; sort_order?: number }) =>
    client.post<AuditItem>('/audit-check/items', body),
  update: (id: number, body: { name?: string; description?: string; sort_order?: number }) =>
    client.put<AuditItem>(`/audit-check/items/${id}`, body),
  toggle: (id: number) => client.patch<AuditItem>(`/audit-check/items/${id}/toggle`),
  remove: (id: number) => client.delete(`/audit-check/items/${id}`),
}

// ── 期別 ──────────────────────────────────────────────────────────────────
export const periodsApi = {
  list: (year?: number) =>
    client.get<Period[]>('/audit-check/periods', { params: year ? { year } : undefined }),
  create: (body: { period: string; title?: string; goal_major?: number; goal_minor?: number; note?: string; review_counts_in_score?: boolean }) =>
    client.post<Period>('/audit-check/periods', body),
  update: (id: number, body: Record<string, unknown>) =>
    client.put<Period>(`/audit-check/periods/${id}`, body),
  remove: (id: number) => client.delete(`/audit-check/periods/${id}`),
}

// ── 稽核單 ────────────────────────────────────────────────────────────────
export const sheetsApi = {
  get: (id: number) => client.get<SheetDetail>(`/audit-check/sheets/${id}`),
  create: (body: {
    period_id: number
    company_id: number
    audited_on?: string | null
    audited_session?: string | null
    executor_label?: string | null
    department_ids: number[]
    items: SheetItemSpec[]
    carry_over?: boolean
  }) => client.post<SheetDetail>('/audit-check/sheets', body),
  update: (id: number, body: Record<string, unknown>) =>
    client.put<SheetDetail>(`/audit-check/sheets/${id}`, body),
  updateLayout: (id: number, body: { department_ids: number[]; items: SheetItemSpec[] }) =>
    client.put<SheetDetail>(`/audit-check/sheets/${id}/layout`, body),
  remove: (id: number) => client.delete(`/audit-check/sheets/${id}`),

  upsertCell: (id: number, body: {
    sheet_item_id: number
    sheet_department_id: number
    comment?: string | null
    result_code?: string | null
  }) => client.put<SheetDetail>(`/audit-check/sheets/${id}/cells`, body),

  upsertCellsBulk: (id: number, cells: {
    sheet_item_id: number
    sheet_department_id: number
    comment?: string | null
    result_code?: string | null
  }[]) => client.put<SheetDetail>(`/audit-check/sheets/${id}/cells/bulk`, { cells }),

  upsertReview: (id: number, sheetDepartmentId: number, body: Record<string, unknown>) =>
    client.put<SheetDetail>(`/audit-check/sheets/${id}/reviews/${sheetDepartmentId}`, body),

  upsertDeficiency: (id: number, sheetDepartmentId: number, text: string | null) =>
    client.put<SheetDetail>(
      `/audit-check/sheets/${id}/departments/${sheetDepartmentId}/deficiency`,
      { deficiency_override: text },
    ),

  /** 完整路徑（給 downloadFile 用，不經過 client.baseURL） */
  exportUrl: (id: number) => `/api/v1/audit-check/sheets/${id}/export`,
}

// ── 統計 ──────────────────────────────────────────────────────────────────
export const statisticsApi = {
  get: (year: number, companyId?: number) =>
    client.get<Statistics>('/audit-check/statistics', {
      params: { year, ...(companyId != null ? { company_id: companyId } : {}) },
    }),
  flagged: (params: {
    period_id?: number
    company_id?: number
    department_id?: number
    only_failing?: boolean
  }) => client.get<FlaggedRow[]>('/audit-check/statistics/flagged', { params }),
  /** 完整路徑（給 downloadFile 用） */
  exportUrl: (year: number) => `/api/v1/audit-check/statistics/export?year=${year}`,
}

// ── 手機版 ────────────────────────────────────────────────────────────────
export const mobileApi = {
  sheets: (limit = 12) =>
    client.get<MobileSheetRow[]>('/audit-check/my-sheets', { params: { limit } }),
}

File: backend/app/schemas/audit_check.py
"""
稽核檢查 — Pydantic Schemas
對應 app/models/audit_check.py、app/routers/audit_check.py
"""
from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, Field


# ── 判定類型 ───────────────────────────────────────────────────────────────
class ResultTypeBase(BaseModel):
    label: str = Field(..., max_length=50)
    color: str = Field("#000000", max_length=20)
    counts_as_pass: bool = True
    include_in_summary: bool = False
    is_default: bool = False
    sort_order: int = 0


class ResultTypeCreate(ResultTypeBase):
    code: str = Field(..., max_length=20, pattern=r"^[a-z][a-z0-9_]{1,19}$")


class ResultTypeUpdate(BaseModel):
    label: Optional[str] = Field(None, max_length=50)
    color: Optional[str] = Field(None, max_length=20)
    counts_as_pass: Optional[bool] = None
    include_in_summary: Optional[bool] = None
    is_default: Optional[bool] = None
    sort_order: Optional[int] = None


class ResultTypeOut(ResultTypeBase):
    id: int
    code: str
    is_system: bool
    is_active: bool

    class Config:
        from_attributes = True


# ── 檢查項主檔 ─────────────────────────────────────────────────────────────
class ItemCreate(BaseModel):
    name: str = Field(..., max_length=200)
    parent_id: Optional[int] = None
    description: Optional[str] = None
    sort_order: int = 0


class ItemUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = None
    sort_order: Optional[int] = None


class ItemOut(BaseModel):
    id: int
    parent_id: Optional[int]
    name: str
    description: Optional[str]
    sort_order: int
    is_active: bool
    in_use: bool = Field(False, description="是否已被任一期稽核單引用（True 則不可改名／刪除）")
    children: List["ItemOut"] = []

    class Config:
        from_attributes = True


ItemOut.model_rebuild()


# ── 期別 ───────────────────────────────────────────────────────────────────
class PeriodCreate(BaseModel):
    period: str = Field(..., pattern=r"^\d{4}-\d{2}$")
    title: str = "財#3系統建置稽核"
    goal_major: int = 3
    goal_minor: int = 3
    note: Optional[str] = None
    review_counts_in_score: bool = True


class PeriodUpdate(BaseModel):
    title: Optional[str] = None
    goal_major: Optional[int] = None
    goal_minor: Optional[int] = None
    note: Optional[str] = None
    review_counts_in_score: Optional[bool] = None
    status: Optional[str] = None


class SheetSummary(BaseModel):
    id: int
    company_id: int
    company_name: str
    audited_on: Optional[date]
    status: str
    completion_rate: Optional[float]
    completion_label: str


class PeriodOut(BaseModel):
    id: int
    period: str
    title: str
    goal_major: int
    goal_minor: int
    note: Optional[str]
    review_counts_in_score: bool = True
    status: str
    created_at: datetime
    sheets: List[SheetSummary] = []

    class Config:
        from_attributes = True


# ── 稽核單 ─────────────────────────────────────────────────────────────────
class SheetItemSpec(BaseModel):
    """建立／調整稽核單時，一列檢查項的規格"""
    item_id: int
    scope_note: Optional[str] = None
    target_department_ids: List[int] = []


class SheetCreate(BaseModel):
    period_id: int
    company_id: int
    audited_on: Optional[date] = None
    audited_session: Optional[str] = None
    executor_label: Optional[str] = None
    department_ids: List[int] = []
    items: List[SheetItemSpec] = []
    carry_over: bool = Field(True, description="是否自動帶入上一期未達標項目為覆核待補正")


class SheetUpdate(BaseModel):
    audited_on: Optional[date] = None
    audited_session: Optional[str] = None
    executor_label: Optional[str] = None
    completion_adjust: Optional[int] = None
    remark: Optional[str] = None
    status: Optional[str] = None


class SheetLayoutUpdate(BaseModel):
    department_ids: List[int]
    items: List[SheetItemSpec]


class SheetDepartmentOut(BaseModel):
    id: int
    department_id: int
    name: str
    sort_order: int
    deficiency_override: Optional[str]
    deficiency: str = Field("", description="缺失列最終顯示值（覆寫優先，否則自動彙整）")


class SheetItemOut(BaseModel):
    id: int
    item_id: int
    parent_sheet_item_id: Optional[int]
    level: int
    display_no: Optional[str]
    name: str
    scope_note: Optional[str]
    target_department_ids: List[int]
    sort_order: int


class CellOut(BaseModel):
    sheet_item_id: int
    sheet_department_id: int
    comment: Optional[str]
    result_code: str
    updated_at: Optional[datetime]


class DepartmentScore(BaseModel):
    sheet_department_id: int
    department_id: int
    name: str
    sub_count: int
    pass_count: int
    score: Optional[float]
    score_label: str


class ReviewOut(BaseModel):
    sheet_department_id: int
    source_period: Optional[str]
    pending_text: Optional[str]
    pending_result: str
    result_text: Optional[str]
    result_status: str


class SheetDetail(BaseModel):
    id: int
    period_id: int
    period: str
    title: str
    goal_major: int
    goal_minor: int
    review_counts_in_score: bool
    company_id: int
    company_name: str
    audited_on: Optional[date]
    audited_session: Optional[str]
    audited_label: str
    executor_label: Optional[str]
    completion_adjust: int
    remark: Optional[str]
    status: str
    departments: List[SheetDepartmentOut]
    items: List[SheetItemOut]
    cells: List[CellOut]
    scores: List[DepartmentScore]
    reviews: List[ReviewOut]
    completion_rate: Optional[float]
    completion_label: str
    result_types: List[ResultTypeOut]


# ── 格子寫入 ───────────────────────────────────────────────────────────────
class CellUpsert(BaseModel):
    sheet_item_id: int
    sheet_department_id: int
    comment: Optional[str] = None
    result_code: Optional[str] = None


class CellBulkUpsert(BaseModel):
    cells: List[CellUpsert]


class ReviewUpsert(BaseModel):
    source_period: Optional[str] = None
    pending_text: Optional[str] = None
    pending_result: Optional[str] = None
    result_text: Optional[str] = None
    result_status: Optional[str] = None


class DeficiencyUpsert(BaseModel):
    deficiency_override: Optional[str] = None


# ── 統計 ───────────────────────────────────────────────────────────────────
class StatisticsCell(BaseModel):
    period: str
    score: Optional[float]
    label: str


class StatisticsRow(BaseModel):
    department_id: int
    name: str
    cells: List[StatisticsCell]


class StatisticsBlock(BaseModel):
    company_id: int
    company_name: str
    periods: List[str]
    audited_labels: List[str]
    major_items: List[List[str]]
    rows: List[StatisticsRow]
    completion: List[StatisticsCell]


class StatisticsOut(BaseModel):
    year: int
    note: Optional[str]
    blocks: List[StatisticsBlock]


class AdviceRow(BaseModel):
    company_name: str
    department_name: str
    item_name: str
    result_code: str
    result_label: str
    result_color: str
    comment: str
